compiler/cpp/src/lexer/Lexer.cpp

```diff
@@ -0,0 +1,279 @@
+    reader.expect('<');
+    if (reader.getLookahead() == '=') {
+        reader.expect('=');
+        return constructToken(TOK_LESS_EQUAL);
+    }
+    return constructToken(TOK_LESS);
+}
+
+Token Lexer::consumeDot() {
+    reader.expect('.');
+    return constructToken(TOK_DOT);
+}
+
+Token Lexer::consumeBraceOpen() {
+    reader.expect('{');
+    currentContext = PAYLOAD;
+    return constructToken(TOK_BRACE_OPEN);
+}
+
+Token Lexer::consumeBraceClose() {
+    reader.expect('}');
+    if (!isLookAheadCommentBlockCloseIdentifier()) currentContext = ARBITRARY;
+    return constructToken(TOK_BRACE_CLOSE);
+}
+
+Token Lexer::consumeIndex() {
+    reader.expect('[');
+    Token indexToken = consumeNumber();
+    reader.expect(']');
+    return constructToken(TOK_INDEX, indexToken.getValue());
+}
+
+Token Lexer::consumeString() {
+    reader.expect('"');
+    std::stringstream stringStr;
+    while(reader.getLookahead() != '"' && reader.getLookahead() != EOF) {
+        if (reader.getLookahead() == '\\') reader.advance();
+        stringStr << (char) reader.getLookahead();
+        reader.advance();
+    }
+    reader.expect('"');
+    return constructToken(TOK_STRING, stringStr.str());
+}
+
+Token Lexer::consumeIdentifierOrKeyword() {
+    std::stringstream identifierStr;
+    identifierStr << (char) reader.getLookahead();
+    reader.advance();
+    while (isalnum(reader.getLookahead())) { // [a-zA-Z0-9]*
+        identifierStr << (char) reader.getLookahead();
+        reader.advance();
+    }
+
+    // Is keyword?
+    std::string identifierName = identifierStr.str();
+    if (identifierName == "if") return constructToken(TOK_IF);
+    if (identifierName == "has") return constructToken(TOK_HAS);
+    if (identifierName == "not") return constructToken(TOK_NOT);
+    if (identifierName == "true") return constructToken(TOK_TRUE);
+    if (identifierName == "false") return constructToken(TOK_FALSE);
+
+    return constructToken(TOK_IDENTIFIER, identifierName);
+}
+
+Token Lexer::consumeNumber() {
+    std::stringstream numStr;
+    do {
+        numStr << (char) reader.getLookahead();
+        reader.advance();
+    } while (isdigit(reader.getLookahead()));
+    return constructToken(TOK_NUMBER, numStr.str());
+}
+
+Token Lexer::consumeCommentLineIdentifier() {
+    reader.expectMultiple(commentLineIdentifier);
+    currentContext = SECTION;
+    return constructToken(TOK_COM_LINE_IDEN);
+}
+
+Token Lexer::consumeCommentBlockOpenIdentifier() {
+    reader.expectMultiple(commentBlockOpenIdentifier);
+    currentContext = SECTION;
+    return constructToken(TOK_COM_BLOCK_IDEN_OPEN);
+}
+
+Token Lexer::consumeCommentBlockCloseIdentifier() {
+    reader.expectMultiple(commentBlockCloseIdentifier);
+    currentContext = ARBITRARY;
+    return constructToken(TOK_COM_BLOCK_IDEN_CLOSE);
+}
+
+Token Lexer::consumeCommentPayloadIdentifier() {
+    reader.expectMultiple(commentPayloadIdentifier);
+    return constructToken(TOK_COM_IDEN_PAYLOAD);
```

Comment:
Although it also works without it, I'd recommend to set `currentContext = SECTION;` here explictly. Will also save some a bit of our preciuos run time ;)

Reply:
Potental bug in your code, see my review in your code.

Reply:
yeah true, fixed this in my code